### DISCUSSION TITLE
List of implemented services

### DISCUSSION THREAD

**Discussion by siGrau** [FreeOpcUa/python-opcua#1226] [General]
Hi,

I´m wondering if there is a list of OPC UA Services, which are implemented in python-opcua.

Thanks in advance

**Comment by AndreasHeine** [comment]
https://github.com/FreeOpcUa/python-opcua/blob/master/README.md

Client

What works:

    connection to server, opening channel, session
    browsing and reading attributes value
    getting nodes by path and nodeids
    creating subscriptions
    subscribing to items for data change
    subscribing to events
    adding nodes
    method call
    user and password
    history read
    login with certificate
    communication encryption
    removing nodes

Tested servers: freeopcua C++, freeopcua Python, prosys, kepware, beckhoff, winCC, B&R, …

Not implemented yet:

    localized text feature
    XML protocol
    UDP
    maybe automatic reconnection...

Server

What works:

    creating channel and sessions
    read/set attributes and browse
    getting nodes by path and nodeids
    autogenerate address space from spec
    adding nodes to address space
    datachange events
    events
    methods
    basic user implementation (one existing user called admin, which can be disabled, all others are read only)
    encryption
    certificate handling
    removing nodes
    history support for data change and events

Tested clients: freeopcua C++, freeopcua Python, uaexpert, prosys, quickopc

Not yet implemented:

    more high level solution to create custom structures
    UDP
    session restore
    alarms
    XML protocol
    views
    localized text features
    better security model with users and password

**Comment by zerox1212** [comment]
Come on now. It takes 2 seconds to check the readme on the main page of the repository.
